pkg/knowledge_base2/resource_template.go

```diff
@@ -38,33 +34,42 @@ type (
 
 		NoIac bool `json:"no_iac" yaml:"no_iac"`
 
-		SanitizeNameTmpl string `yaml:"sanitize_name"`
+		SanitizeNameTmpl SanitizeTmpl `yaml:"sanitize_name"`
 	}
 
-	Properties map[string]*Property
-
-	Property struct {
-		Name string `json:"name" yaml:"name"`
-		// Type defines the type of the property
-		Type string `json:"type" yaml:"type"`
-
-		Namespace bool `json:"namespace" yaml:"namespace"`
-
-		DefaultValue any `json:"default_value" yaml:"default_value"`
-
-		Required bool `json:"required" yaml:"required"`
-
+	PropertyDetails struct {
+		Name string
+		// DefaultValue has to be any because it may be a template and it may be a value of the correct type
+		Namespace bool `yaml:"namespace"`
+		// Required defines if the property is required
+		Required bool
+		// ConfigurationDisabled defines if the property is allowed to be configured by the user
 		ConfigurationDisabled bool `json:"configuration_disabled" yaml:"configuration_disabled"`
-
+		// DeployTime defines if the property is only available at deploy time
 		DeployTime bool `json:"deploy_time" yaml:"deploy_time"`
-
-		OperationalRule *OperationalRule `json:"operational_rule" yaml:"operational_rule"`
-
-		Properties map[string]*Property `json:"properties" yaml:"properties"`
-
+		// OperationalRule defines a rule that is executed at runtime to determine the value of the property
+		OperationalRule *PropertyRule `json:"operational_rule" yaml:"operational_rule"`
+		// Path is the path to the property in the resource
 		Path string `json:"-" yaml:"-"`
 	}
 
+	Property interface {
+		SetProperty(resource *construct.Resource, value any) error
+		AppendProperty(resource *construct.Resource, value any) error
+		RemoveProperty(resource *construct.Resource, value any) error
+		Details() *PropertyDetails
+		Clone() Property
+		Type() string
+		GetDefaultValue(ctx DynamicValueContext, data DynamicValueData) (any, error)
+		Validate(value any, properties construct.Properties) error
+		SubProperties() map[string]Property
```

Comment:
This can return a `Properties` now that this type's not tied to the yaml parsing, right?